core/src/main/java/org/apache/accumulo/core/clientImpl/bulk/ConcurrentKeyExtentCache.java

```diff
@@ -88,9 +88,10 @@ protected void updateCache(KeyExtent e) {
 
   @VisibleForTesting
   protected Stream<KeyExtent> lookupExtents(Text row) {
-    return TabletsMetadata.builder(ctx).forTable(tableId).overlapping(row, true, null)
-        .checkConsistency().fetch(PREV_ROW).build().stream().limit(100)
-        .map(TabletMetadata::getExtent);
+    try (TabletsMetadata tabletsMetadata = TabletsMetadata.builder(ctx).forTable(tableId)
+        .overlapping(row, true, null).checkConsistency().fetch(PREV_ROW).build()) {
+      return tabletsMetadata.stream().limit(100).map(TabletMetadata::getExtent);
```

Comment:
This is returning a stream, so the TabletsMetadata object can not be closed.  Could add an onClose call to the stream to close tabletsMetadata.  The calling code would need to close the stream when its done

Reply:
I ended up making changes to this case along the same lines as your suggestion but am thinking that passively relying on other code to close the stream seems fragile and there is no guarantee that it will happen. I am looking into these cases and trying to see if they can be refactored to return a TabletsMetadata object directly so that the calling code is more likely to deal with closing the resource but is seems like lots of places are going to have to use the `onClose()`.

This is just an observation but seems like a good argument against patterns that do not enforce closing of resources like this.

Reply:
I think returning an Iterator might be okay since we can add a step to make sure the iterator is created from a collection created from the resource rather than directly off of the resource but it seems more hacky to do that with a stream, i.e. converting from a stream to a collection then back to a stream in order to de-couple from the underlying resource.